### PR TITLE
feat(deploy): apply publishConfig to all packages during deploy

### PR DESCRIPTION
**Pull request by JacobLey**
When deploying packages, the package.json of the deployed package (as well as any other locally defined dependencies) should be treated as if it published, and mutate the package.json according to `publishConfig` and local `workspace:` dependencies.

Issue: https://github.com/pnpm/pnpm/issues/6693

### CODE REVIEW

File: releasing/plugin-commands-deploy/test/deploy.test.ts
Comment (JacobLey) [inline releasing/plugin-commands-deploy/test/deploy.test.ts:63]: In order to `createExportableManifest` we need metadata only available after an install.

Presently, `deploy` does not actually require an install previously.

This _could_ be considered a breaking change. Although I would assume most users already have the relevant packages installed, and expected this behavior. This is inline with how commands like `publish` work.

`deploy` could be considered a "local" `publish`.

File: releasing/plugin-commands-deploy/test/deploy.test.ts
Comment (JacobLey) [inline releasing/plugin-commands-deploy/test/deploy.test.ts:65]: A little bit of a hack for testing purposes... an assertion below checks that this file _does not exist_ (as a side effect of the `install` called internally to `deploy`) and wanted to maintain that assertion.

File: fs/indexed-pkg-importer/src/importIndexedDir.ts
Comment (zkochan) [inline fs/indexed-pkg-importer/src/importIndexedDir.ts]: This is going to be used only by the deploy command, so I don't see any reason to put this code to importIndexedDir. Just place it in `deploy.ts`

**Review by zkochan — CHANGES_REQUESTED** [review]

File: fs/indexed-pkg-importer/src/importIndexedDir.ts
Comment (JacobLey) [reply fs/indexed-pkg-importer/src/importIndexedDir.ts]: Updated to put all `createExportableManifest` logic in `deploy.ts`.

Removed all changes to other packages (adding `createExportableManifest` option)

**Comment by JacobLey** [comment]
Build is currently _only_ failing on ubuntu+v16 on this test: https://github.com/pnpm/pnpm/blob/main/pnpm/test/filterProd.test.ts#L9

This PR does not touch filtering, install, or recursive execution, so I would be surprised if I explicitly broke something.

Will defer to maintainer for guidance if there is something I can do to fix? Or perhaps just retry the test (flaky)?

**Comment by zkochan** [comment]
The issue with this solution is that now `createExportableManifest` will be executed on all the dependencies. Even though we only need to run it on dependencies linked from local directories.

I suggest a different solution. Make the change in [directory-fetcher](https://github.com/pnpm/pnpm/blob/main/fetching/directory-fetcher/src/index.ts). It create an index file that is passed to the linker. You can write the "exportable manifest" into `node_modules/.pnpm/package.json` and provide the path to it to the files index map. This way the `package.json` file that you create will be linked, not the original one.

**Comment by JacobLey** [comment]
> will be executed on all the dependencies

Agreed, should only be executed on local packages.

Looking into `directory-fetcher`, but we are back to the original issue that we will need to start passing some flag around to other packages for this single use case. Which I am ok implementing, but calling it out.

Another possible solution is to look up the list of all packages in the workspace pre-install (via something like `findWorkspacePackages`). Then reference that list for each hook and check "is this package a local workspace package". If true -> `createExportableManifest`.
There may even be an existing utility that does that check?

> write the "exportable manifest" into `node_modules/.pnpm/package.json`

That sounds safe to me. Do we want to clean up the file after we are done with it?
Such as 
```
process.on('exit', () => fs.rm('node_modules/.pnpm/package.json')
```

**Comment by zkochan** [comment]
> Looking into directory-fetcher, but we are back to the original issue that we will need to start passing some flag around to other packages for this single use case. Which I am ok implementing, but calling it out.

No, I think we can make this the default behaviour for all local directory dependencies. No need to pass options.

**Comment by JacobLey** [comment]
Updated, but not ready to merge.

Tests are now (rightfully) failing on PnP recursion https://github.com/pnpm/pnpm/blob/main/exec/plugin-commands-script-runners/test/exec.e2e.ts#L487

Not _entirely_ sure how I broke that, but there must be some expectation that the package.json link remains unaltered.

File: pkg-manifest/read-project-manifest/src/index.ts
Comment (JacobLey) [inline pkg-manifest/read-project-manifest/src/index.ts]: The existing behavior was to re-write the file in same location it was loaded from. That is unwanted for our use case, so provided this override option.

This "breaks" existing options syntax, although I did not find any usage of the force option... so actually worked in-repo.

If this method is considered part of pnpm's public API though, I can tweak to accept a boolean _or_ this options object for backwards compatibility.

Alternatively, I can use the `writeProjectManifest` function directly in `directory-fetcher`, but I did appreciate that the `manifest.writeProjectManifest` came with sensible defaults

File: fetching/directory-fetcher/src/index.ts
Comment (JacobLey) [inline fetching/directory-fetcher/src/index.ts]: Not sure exactly what behavior depends on it, but would it be preferable to return `exportableManifest` here instead?

FWIW that did not solve my current failing tests anyways.

File: fetching/directory-fetcher/src/index.ts
Comment (JacobLey) [inline fetching/directory-fetcher/src/index.ts]: Here is the use case of `manifestPathOverride`. As mentioned in other comment, could replace this call with raw `writeProjectManifest` (rather than the one built into the safeRead response)

File: fetching/directory-fetcher/src/index.ts
Comment (zkochan) [inline fetching/directory-fetcher/src/index.ts]: Let's actually check if the exportableManifest differs from the original one. If it doesn't differ, there's no. need to write it to the file system and the path to the original file may be used.

File: pkg-manifest/read-project-manifest/src/index.ts
Comment (zkochan) [reply pkg-manifest/read-project-manifest/src/index.ts]: Don't add this option. Just use a function that accepts the path to the manifest for saving the manifest.

**Comment by JacobLey** [comment]
Updated to use separate `write` function, so doesn't alter existing parameters. 
Also check for diff before writing, to hopefully optimize file write.

One (trivial) call out is that the package will not necessarily be formatted the same way as before. Which I guess wasn't ever a guarantee (e.g. multi-newline in JSON file is removed) but now we would also lose comments in a JSON5, or change the indent settings.

**Review by zkochan — APPROVED** [review]

**Pull request merged**
